dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/util/PeriodOffsetUtils.java

```diff
@@ -0,0 +1,231 @@
+
+        for ( DimensionalItemObject item : items )
+        {
+            if ( item.getPeriodOffset() != 0 )
+            {
+                shiftedMap.putAll( addPeriodOffset( periodTypePeriodMap, item.getPeriodOffset() ) );
+            }
+        }
+
+        Set<DimensionalItemObject> dimensionalItemObjects = shiftedMap.uniqueValues();
+        for ( DimensionalItemObject dimensionalItemObject : dimensionalItemObjects )
+        {
+            Period period = (Period) dimensionalItemObject;
+            if ( !periodTypePeriodMap.containsValue( period.getPeriodType().getName(), dimensionalItemObject ) )
+            {
+                periodTypePeriodMap.putValue( period.getPeriodType().getName(), dimensionalItemObject );
+            }
+        }
+
+        return periodTypePeriodMap;
+    }
+
+    /**
+     * Shifts the given Period in the past or future based on the offset value.
+     * 
+     * Example:
+     * 
+     * Period: 202001 , Offset: 1 -> Period: 202002
+     * Period: 2020 , Offset: -1 -> Period: 2019
+     * 
+     * @param period a Period
+     * @param periodOffset a positive or negative integer
+     * @return A Period
+     */
+    public static Period shiftPeriod( Period period, int periodOffset )
+    {
+        if ( periodOffset == 0 )
+        {
+            return period;
+        }
+
+        PeriodType periodType = period.getPeriodType();
+        Period p;
+        if ( periodOffset > 0 )
+        {
+            p = periodType.getNextPeriod( period, periodOffset );
+        }
+        else
+        {
+            p = periodType.getPreviousPeriod( period, periodOffset );
+        }
+
+        p.setShifted( true );
+        return p;
+    }
+
+    /**
+     * Remove Periods from a {@see DataQueryParams} object if these periods have
+     * been added because of an "periodOffset" directive and the DataElement have no
+     * offset specified. This can happen in case of an Indicator, where a numerator
+     * formula is using an offset, and the denominator formula is not.
+     * 
+     * @param params a {@see DataQueryParams} object
+     * @return a {@see DataQueryParams} object
+     */
+    public static DataQueryParams removeOffsetPeriodsIfNotNeeded( DataQueryParams params )
+    {
+        final List<DimensionalItemObject> dimensionalItemObjects = params.getDataElements();
+
+        final boolean hasOffset = dimensionalItemObjects.stream().filter( dio -> dio.getDimensionItemType() != null )
+            .filter( dio -> dio.getDimensionItemType().equals( DimensionItemType.DATA_ELEMENT ) )
+            .anyMatch( dio -> dio.getPeriodOffset() != 0 );
+
+        if ( !hasOffset )
+        {
+            final List<DimensionalItemObject> nonShiftedPeriods = params.getPeriods().stream()
+                .filter( dio -> (!((Period) dio).isShifted()) ).collect( Collectors.toList() );
+
+            // TODO is there a better way to "replace" periods?
```

Comment:
Yes there is. In the builder, you can use `Builder.withPeriods(..)`.

So something like:

```
DataQueryParams.newBuilder( params )
    .withPeriods( nonShiftedPeriods )
    .build();
```